google/auth/_default.py

```diff
@@ -0,0 +1,295 @@
+# ID.
+_CREDENTIALS_ENV = 'GOOGLE_APPLICATION_CREDENTIALS'
+_PROJECT_ENV = 'GCLOUD_PROJECT'
+
+# Valid types accepted for file-based credentials.
+_AUTHORIZED_USER_TYPE = 'authorized_user'
+_SERVICE_ACCOUNT_TYPE = 'service_account'
+_VALID_TYPES = (_AUTHORIZED_USER_TYPE, _SERVICE_ACCOUNT_TYPE)
+
+# The Google OAuth 2.0 token endpoint. Used for authorized user credentials.
+_GOOGLE_OAUTH2_TOKEN_ENDPOINT = 'https://accounts.google.com/o/oauth2/token'
+
+# The ~/.config subdirectory containing gcloud credentials.
+_CLOUDSDK_CONFIG_DIRECTORY = 'gcloud'
+# The environment variable name which can replace ~/.config if set.
+_CLOUDSDK_CONFIG_ENV = 'CLOUDSDK_CONFIG'
+# The name of the file in the Cloud SDK config that contains default
+# credentials.
+_CLOUDSDK_CREDENTIALS_FILENAME = 'application_default_credentials.json'
+# The name of the file in the Cloud SDK config that contains the
+# active configuration.
+_CLOUDSDK_ACTIVE_CONFIG_FILENAME = os.path.join(
+    'configurations', 'config_default')
+# The config section and key for the project ID in the cloud SDK config.
+_CLOUDSDK_PROJECT_CONFIG_SECTION = 'core'
+_CLOUDSDK_PROJECT_CONFIG_KEY = 'project'
+
+# Help message when no credentials can be found.
+_HELP_MESSAGE = (
+    'Could not automatically determine credentials. Please set {env} or '
+    'explicitly create credential and re-run the application. For more '
+    'information, please see https://developers.google.com/accounts/docs'
+    '/application-default-credentials.'.format(env=_CREDENTIALS_ENV))
+
+
+def _load_credentials_from_file(filename):
+    """Loads credentials from a file.
+
+    The credentials file must be a service account key or stored authorized
+    user credentials.
+
+    Args:
+        filename (str): The full path to the credentials file.
+
+    Returns:
+        Tuple[google.auth.credentials.Credentials, Optional[str]]: Loaded
+            credentials and the project ID. Authorized user credentials do not
+            have the project ID information.
+
+    Raises:
+        google.auth.exceptions.DefaultCredentialsError: if the file is in the
+            wrong format.
+    """
+    with io.open(filename, 'r') as file_obj:
+        try:
+            info = json.load(file_obj)
+        except ValueError as exc:
+            raise exceptions.DefaultCredentialsError(
+                'File {} is not a valid json file.'.format(filename), exc)
+
+    # The type key should indicate that the file is either a service account
+    # credentials file or an authorized user credentials file.
+    credential_type = info.get('type')
+
+    if credential_type == _AUTHORIZED_USER_TYPE:
+        credentials = google.oauth2.credentials.Credentials(
+            None,
+            refresh_token=info['refresh_token'],
+            token_uri=_GOOGLE_OAUTH2_TOKEN_ENDPOINT,
+            client_id=info['client_id'],
+            client_secret=info['client_secret'])
```

Comment:
(It may seem weird because service_account.json is also google-specific, but it is a public file format, while the authorized_user format is not public but is an implementation detail of the cloud sdk, so if it should go anywhere it should be grouped with the cloud sdk specific code).